File: Algos/DFS.cpp
#include <algorithm>
#include <iostream>
#include <stack>
#include <vector>

void
dfs(int vertex, std::vector<std::vector<int> > &adj, std::vector<bool> &visited) {
    if (visited[vertex]) {
        return;
    }
    visited[vertex] = true;
    std::cout << vertex << " ";
    for (const auto &v : adj[vertex]) {
        dfs(v, adj, visited);
    }
}

void
iterative_dfs(int vertex, std::vector<std::vector<int> > &adj, std::vector<bool> &visited) {
    for (auto &&v : visited) {
        v = false;
    }
    std::stack<int> stack;
    visited[vertex] = true;
    stack.push(vertex);

    while (!stack.empty()) {
        int curr = stack.top();
        std::cout << curr << " ";
        stack.pop();
        for (int i = (int) adj[curr].size() - 1; i >= 0; i--) {
            auto &v = adj[curr][i];
            if (!visited[v]) {
                visited[v] = true;
                stack.push(v);
            }
        }
    }
}

int
main() {
    std::ios_base::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);

    int vertices;
    std::cin >> vertices;
    std::vector<bool> visited(vertices + 1, false);
    std::vector<std::vector<int> > adj(vertices + 1, std::vector<int>());
    int edges;
    std::cin >> edges;

    for (int i = 0; i < edges; i++) {
        int x, y;
        std::cin >> x >> y;
        adj[x].push_back(y);
        // adj[y].push_back(x);
    }
    dfs(1, adj, visited);
    std::cout << "\n";
    iterative_dfs(1, adj, visited);

    return 0;
}

File: Algos/python_versions/dfs.py
def dfs(ver, adj, visited):
    visited[ver] = True
    print(ver, end=" ")
    for v in adj[ver]:
        if visited[v] == False:
            dfs(v, adj, visited)


def iter_dfs(ver, adj, visited):
    stack = []
    stack.append(ver)
    visited[ver] = True
    while len(stack) > 0:
        curr = stack[len(stack) - 1]
        print(curr, end=" ")
        stack.pop()
        visited[curr] = True
        for i in range(len(adj[curr]) - 1, -1, -1):
            if visited[adj[curr][i]] == False:
                stack.append(adj[curr][i])


v = int(input())
visited = [False] * (v + 1)
adj = [[] for _ in range(v + 1)]
edges = int(input())
for i in range(edges):
    inp = input().split()
    x = int(inp[0])
    y = int(inp[1])
    adj[x].append(y)
dfs(1, adj, visited)
print()
visited = [False] * (v + 1)
iter_dfs(1, adj, visited)
